### DISCUSSION TITLE
1:1 Functionality with Openbox for LXQt-Wayland session

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [labwc/labwc#1105] [Ideas]
Hello! I'm one of the contributors to the LXQt project, though only as a themer since I can't code. LXQt has become my go-to desktop, and it's very exciting to see it making progress on a Wayland implementation.

Despite @tsujan's insistence that LXQt is to remain WM-agnostic, I think we're kidding ourselves by not recognizing that basically everyone uses it with Openbox on X11. To that end, and in parallel with @stefonarch's lxqt-wayland repo, I would very much like to see LabWC become the "soft default" on Wayland as Openbox is on X11.

However, we have a problem: LabWC doesn't implement some of the theming settings that Openbox does, and the README specifically says things like gradients are not planned to be supported.

My question is, why not? It can't be _that_ difficult to add support, and given that LabWC:Wayland :: Openbox:X11 (more or less) it doesn't make sense to be missing such a central feature, doubly so for something that will be seen by most end users as "the same thing but Wayland."

So, this is more of a feature request than anything (to please bring LabWC's theming to 1:1 compatibility with Openbox) for the sake of future cohesion and adaptability of the LXQt desktop. The latest release has gone above and beyond by adding support for SVG window buttons, so it's not as if this is an impossible ask.

If I could code I'd do this myself. But even if I could, this runs the risk of creating a schism, which would demoralize basically everyone on the LabWC team, especially if a fork with full theme support becomes the de facto standard version of LabWC. If this support is added it really should be done in the main repo here, and with full consent and support of the master developers.

**Comment by johanmalm** [comment]
Thanks for reaching out :smile:

You making a compelling point and I don't feel strongly about gradients, icons and drop-shadows. I think in hind-sight that stance - which has sort of stuck - was probably a way of mentally coping with the size of the task that writing a compositor comparable to Openbox constitutes. 

We've been going steadily (albeit slowly) for over three years and I estimate that there is another year's worth of effort before I want to think too much about extending scope (in terms of what I personally set out to do). 

I'll chat to the other core devs to see if we can reach consensus toning down the README.

As long as we never compromise production quality within current scope, stability and good/extendable code design, etc, I'm good.

**Reply by tsujan** [reply]
> I don't feel strongly about gradients, icons and drop-shadows.

IMVHO, the title-bar gradient isn't in the same category as shadows.

I saw the closed report of a user who even talked of fade-in/fade-out and blur effects. I really doubt that Openbox users miss shadows (which didn't and couldn't exist in Openbox), let alone the blur effect, which can easily be a resource hog. As someone who has used X11 compositors with shadows and blur effect for more than a decade, I didn't miss them when I started to use LabWC, because I saw how stable, snappy and reliable LabWC was. Who wants to sacrifice that for shadows, animations or blurring? ;)

But I'm afraid I can't say the same thing about a title-bar gradient — especially when I see that LabWC supports nice rounded corners.

**Reply by Consolatis** [reply]
We changed our stance on some of the things.
- See #1125 for a discussion and the outcome.

Basically:
- Icons + gradients are fine (although not yet implemented)
- Shadows and blur only if they
  - don't drag down performance
  - can be disabled (and are disabled by default on software rendering for example)
  - don't add enormous amounts of complexity to the codebase

**Reply by tsujan** [reply]
Thanks!

> Shadows and blur only if they don't drag down performance,

As far as I've seen, blurring *does* drag down performance, whether under X11 or on Wayland (Wayfire). Actually, the only X11 compositor which prevents that *as far as possible* is KWin, but KWin's implementation is based on requests made by codes of Qt styles and/or apps (among other things); it doesn't blur every translucent region.

**Reply by Consolatis** [reply]
I don't think any of the core devs will implement shadows anyway. This is basically more of a criteria when somebody else comes along and sends a PR with an implementation.

**Reply by tsujan** [reply]
Not being familiar with `wlroots` and not intending to bother you with my code questions, I just wanted to ask one:

Am I right supposing that `wlroots` doesn't support gradients, and that adding a gradient requires adding a properly resizable Cairo layer over the title-bar rectangle?

**Reply by Consolatis** [reply]
wlroots itself doesn't support any cairo stuff at all.
It does support a very basic `wlr_scene_rect` which is basically a unique color rect that can be resized / recolored. We are using that for borders and various backgrounds currently (like for the titlebar or menu).

wlroots also supports pixel buffers, those we use for the A-Tab OSD or the window title itself + rounded corners + window icons.
So to implement gradients we'd have to replace the current `wlr_scene_rect` titlebar background with a `wlr_scene_buffer` which we draw into using cairo. We should also hook that up to our "autoscaling buffer" so it gets (re)rendered using the current output scale when moving the window to another output with a different scale. Otherwise it ends up looking blurry. That isn't currently necessary for the single color rectangle we are using as it scales just fine.

**Reply by johanmalm** [reply]
I think the main obstacle is that the ‘titlebar’ is currently made up of two corner pieces and a middle piece. 
As gradients can go in all directions, I suspect we’d have to refactor the above  so that we just have one titlebar wlr_scene_buffer

**Reply by Consolatis** [reply]
Hm.. right. That is a good point. I am not really that familiar with the details of Openbox theming, If the gradients indeed include the corner pieces as well that would mean that we basically have to rerender all 3 parts on every width change.

**Reply by tsujan** [reply]
Thanks for the replies!

A gradient from top to bottom should be enough for almost all users (assuming that Openbox supported more — I'm not familiar with Openbox's theming either).

**Reply by Consolatis** [reply]
Just a random thought: We could cheat a little here:
- keep splitting the corners and the middle part
- when resizing
  - if it gets wider: re-render the middle part
  - if it gets narrower: crop the buffer (I think this works for wlr_scene_buffer but we don't do that anywhere yet)
- when ending the resize, render everything correctly and remove any buffer cropping

That should prevent unnecessary lags during resize and still look somewhat decent. When the resize is done everything is back to normal.

Edit:
Another thing that we could do is render the corners as a bitmask once when loading the config. When then rendering the individual corners (or a single buffer which contains the whole titlebar background, including the corners) we use the bitmasks to "stamp out" the transparent pixels on the edges.

**Reply by johanmalm** [reply]
For ref: http://openbox.org/wiki/Help:Themes#Gradients
Openbox supports many gradients other than `vertical`

**Reply by tsujan** [reply]
I'd get rid of those variations and just treat "color" and "colorTo" as a linear gradient from top to bottom; just my opinion.

**Reply by Consolatis** [reply]
It certainly would make stuff easier. We might be even able to keep the separation of the corners and the "middle" part that way:
Radial gradient for the corners + linear gradient for the middle part should (hopefully) connect them together seamlessly.

**Reply by tsujan** [reply]
> Radial gradient for the corners

Why radial? I tested with a linear gradient, and it looked quite natural (as it did with Openbox).

**Reply by tsujan** [reply]
Let me "confess" my use case ;)

There are Kvantum themes that have gradients in their toolbars/menubars. With appropriate Openbox themes, those gradients were extended to Openbox's title-bar, creating a seamless total gradient.

**Reply by Consolatis** [reply]
Do you have some screenshots?

**Reply by tsujan** [reply]
I didn't take a screenshot before reversing the changes (because I wasn't able to do anything with the middle part), but it looked as natural as KWin, Openbox, or this (an Emerald theme from old times):

![theme screenshot](https://github.com/labwc/labwc/assets/981076/4314930c-5b10-4d00-b7a0-95559b52d32d)

**Reply by Consolatis** [reply]
I see. So just vertical linear gradient for all parts and then "cut-off" rounded corners.

**Reply by tsujan** [reply]
Yes :) Even KDE's "Breeze" title-bar style doesn't do anything more than that (although I had to correct its suboptimal gradient in [BreezeEnhanced](https://github.com/tsujan/BreezeEnhanced)).

**Reply by AzumaHazuki** [reply]
This is kind of missing the point...if LabWC is to be Openbox's Wayland doppelganger in the context of LXQt, it has to render themes identically.

I don't know how different it is to code these things in Wayland vs Openbox, but it _is_ F/OSS, so the source is available right? Maybe an easier way to implement the titlebar and buttons is to have them each be a generic "drawable rectangle" object of some sort, which takes the borders, raised/flat/sunken, horizontal/vertical/radial/crossdiagonal etc. gradient properties? And would perform the proper action on click - raise/focus if it's the titlebar or label, close/min/max/restore/stick if it's the respective button, and so on.

**Reply by tsujan** [reply]
> if LabWC is to be Openbox's Wayland doppelganger

I've assumed it isn't, based on what is said explicitly in README: "This does not mean that labwc is an openbox clone…"

**Reply by johanmalm** [reply]
@AzumaHazuki I see no issue with rendering as Openbox. Being `theme` compatible with Openbox is important to me. `labwc` supports additional stuff, but I wouldn't want to compromise openbox theme spec. We're just not openbox theme-spec complete yet (gradients and hover icons being two notable gaps).

Implementing gradients is not a simple patch. If we didn't support rounded corner in that way that we do, it probably would have been easy :smile:

We discussed extensively on IRC a few weeks ago and decided to try a simple implementation with some vertical gradients to make sure it's not too CPU intensive for very modest hardware. If it is, we'll have to consider a more sophisticated approach.

In brief, the main obstacles making this take some time are:
(a) It's currently not a priority of core devs
(b) Corners are currently rendered separately from the rest of the titlebar. For non-vertical gradients this is a pretty awkward starting point, so we need to do some refactoring to overcome that.

We try to accommodate requests - but it's not always at the pace people want it. When that is the case, my advice is to either be patient or find someone who can code and ask them to implement it. We'll always merge sensible patches unless they really complicate things or make maintainability hard in some other way.

**Reply by redtide** [reply]
> A gradient from top to bottom should be enough for almost all users (assuming that Openbox supported more — I'm not familiar with Openbox's theming either).

I'm using the Onyx-Cytrus theme and it uses what I call a "glossy gradient", which is composed by 2 gradients: top to middle and middle to bottom; I played with this thing using wxGTK years ago.
